src/lex.cpp

```diff
@@ -849,15 +850,22 @@ char8* lexer::parse_decimal_digits_and_underscores(char8* input) noexcept {
           input += 1;
         }
 
-        this->error_reporter_->report(
+        if (is_digit(*input)) {
+          this->error_reporter_->report(
             error_number_literal_contains_consecutive_underscores{
-                source_code_span(garbage_begin, input)});
+              source_code_span(garbage_begin, input)});
+        } else {
+          this->error_reporter_->report(
+            error_number_literal_contains_trailing_underscores{
+              source_code_span(garbage_begin, input)});
+        }
       }
     }
   }
   if (has_trailing_underscore == true) {
-    // TODO error_reporter: SyntaxError: Numeric separators are not allowed at
-    // the end of numeric literals
+    this->error_reporter_->report(
+        error_number_literal_contains_trailing_underscores{
+            source_code_span(garbage_begin, input)});
```

Comment:
Blocking: Clang reports an [error when compiling this code](https://github.com/strager/quick-lint-js/runs/1228354630):

```
[ 49%] Building CXX object src/CMakeFiles/quick-lint-js-lib.dir/lex.cpp.o
/Users/runner/work/quick-lint-js/quick-lint-js/src/lex.cpp:875:30: error: variable 'garbage_begin' may be uninitialized when used here [-Werror,-Wconditional-uninitialized]
            source_code_span(garbage_begin, input)});
                             ^~~~~~~~~~~~~
/Users/runner/work/quick-lint-js/quick-lint-js/src/lex.cpp:845:29: note: initialize the variable 'garbage_begin' to silence this warning
  const char8* garbage_begin;
                            ^
                             = nullptr
1 error generated.
```

[GCC reports a similar error.](https://github.com/strager/quick-lint-js/runs/1228354758) These errors cause CI builds to fail. (CI builds are misconfigured on pull requests (my fault), so I didn't see this issue until I tried merging.)

I think this diagnostic is a false positive (i.e. there's no bug in your code), but we should fix the diagnostic anyway.

Idea: Instead of having two variables (`garbage_begin` and `has_trailing_underscore`), we could use a single variable. I'm thinking `garbage_begin` can be `nullptr` initially, and we check `if (garbage_begin) {` on line 365.

Reply:
I do not know have an error with my compiler (`gcc (Debian 8.3.0-6)`) but I agree with you on the false positive.
Indeed, `garbage_begin` can be uninitialized after the loop.
But if we enter the if after the loop, `garbage_begin` had been set since it is set before `has_trailing_underscore` which is the `if` condition.

I think there is a problem with your solution.
Indeed, if we only condition the outer loop `if` with `garbage_begin` a number like 123__456 will print both consecutive and trailing underscores error.
To make the compiler happy and having all cases well covered I add `garbage_begin != nullptr` to the condition of outer loop if.